### PR TITLE
ENH: CIFTI carpetplot

### PR DESCRIPTION
**Pull request by mgxd**
This addresses poldracklab/fmriprep#2026 with the idea of replacing the current volumetric carpetplot with a surface/voxel hybrid.

An example rendered image:
Screen Shot 2020-03-27 at 6 10 12 PM

This includes 4 groupings: left and right cortices (dark/light blue respectively), subcortical structures (orange), and cerebellum (green).

### CODE REVIEW

**Comment by mgxd** [comment]
note that the latest commit adds an fmriprep generated CIFTI from ds005 (11.7MB)

File: docs/api.rst [hunk 1]
```diff
@@ -7,6 +7,7 @@ Information on specific functions, classes, and methods.
    :glob:
 
    api/niworkflows.anat
+   api/niworkflows.cli
```
Comment (oesteban) [inline docs/api.rst:10]: we should address this problem separately and fix master first (sorry, I introduced this with #488)

File: niworkflows/viz/plots.py [hunk 2]
```diff
@@ -24,33 +25,48 @@
...
-class fMRIPlot(object):
+class fMRIPlot:
     """
     Generates the fMRI Summary Plot
     """
-    __slots__ = ['func_file', 'mask_data',
-                 'tr', 'seg_data', 'confounds', 'spikes']
+    __slots__ = ('func_file', 'mask_data', 'tr', 'seg_data', 'confounds', 'spikes', '_cifti')
+
+    def __init__(
```
Comment (oesteban) [inline niworkflows/viz/plots.py:34]: Are there any changes to the function's signature or this is just black's reformatting?

File: niworkflows/viz/plots.py [hunk 3]
```diff
@@ -24,33 +25,48 @@
...
+        units=None,
+        vlines=None,
+        spikes_files=None,
+    ):
+        self._cifti = False
+        try:
+            func_img = nb.cifti2.load(func_file)
+            self._cifti = True
+        except (ImageFileError, HeaderDataError):
+            func_img = nb.load(func_file)
```
Comment (oesteban) [inline niworkflows/viz/plots.py:52]: Does `nb.load()` accept CIFTI2, @effigies ? If so, I guess it'd be more elegant to just load and determine whether this is cifti2 by looking at the instance's class.

File: niworkflows/viz/plots.py [hunk 4]
```diff
@@ -271,6 +290,128 @@ def plot_carpet(img, atlaslabels, detrend=True, nskip=0, size=(950, 800),
...
+        ax0.spines[side].set_color('none')
+        ax0.spines[side].set_visible(False)
+        ax1.spines[side].set_color('none')
+        ax1.spines[side].set_visible(False)
+
+    ax1.yaxis.set_ticks_position('left')
+    ax1.xaxis.set_ticks_position('bottom')
+    ax1.spines["bottom"].set_visible(False)
+    ax1.spines["left"].set_color('none')
+    ax1.spines["left"].set_visible(False)
```
Comment (oesteban) [inline niworkflows/viz/plots.py:404]: I don't love all this duplicated code... Would it be possible to extract this out to a common, private function for both volumetric and cifti2 carpets? In the end the pseudocode of the plotting would be:

```Python
def _carpet(data, segments, lut):
    <the common code for plotting>
    return plot

def plot_carpet(data, datatype='NIFTI'):
    if datatype.upper().startswith("cifti"):
        """calculate segments and color lookup table (lut) for CIFTI"""
    else:
        """calculate segments and color lookup table (lut) for NIfTI"""
    
    return _carpet(data, segments, lut)

File: niworkflows/viz/plots.py [hunk 5]
```diff
@@ -770,3 +911,28 @@ def confounds_correlation_plot(confounds_file, output_file=None, figure=None,
         figure = None
         return output_file
     return [ax0, ax1], gs
+
+
+def _get_tr(img, is_cifti=False):
```
Comment (oesteban) [inline niworkflows/viz/plots.py:916]: Let's go more pythonic EAFP:
```suggestion
def _get_tr(data):
```

File: niworkflows/viz/plots.py [hunk 6]
```diff
@@ -770,3 +911,28 @@ def confounds_correlation_plot(confounds_file, output_file=None, figure=None,
...
+    2.2
+    >>> _get_tr(ci, is_cifti=True)
+    2.0
+    """
+
+    if is_cifti:
+        tr = img.header.matrix.get_index_map(0).series_step
+    else:
+        tr = img.header.get_zooms()[-1]
+    return tr
```
Comment (oesteban) [inline niworkflows/viz/plots.py:938]:
```suggestion
    try:
        return data.header.matrix.get_index_map(0).series_step
    except AttributeError:
        return data.header.get_zooms()[-1]
    raise RuntimeError("Could not extract TR - unknown data structure type")
```

File: niworkflows/viz/plots.py [hunk 7]
```diff
@@ -770,3 +911,28 @@ def confounds_correlation_plot(confounds_file, output_file=None, figure=None,
...
+    >>> ci = nb.load(Path(test_data) /
+    ...    'sub-01_task-mixedgamblestask_run-02_space-fsLR_den-91k_bold.dtseries.nii')
+
+    Examples
+    --------
+    >>> _get_tr(nii)
+    2.2
+    >>> _get_tr(ci, is_cifti=True)
+    2.0
+    """
```
Comment (oesteban) [inline niworkflows/viz/plots.py:932]:
```suggestion
    Examples
    --------
    >>> _get_tr(nb.load(test_data /
    ...    'sub-ds205s03_task-functionallocalizer_run-01_bold_volreg.nii.gz'))
    2.2
    >>> _get_tr(nb.load(test_data /
    ...    'sub-01_task-mixedgamblestask_run-02_space-fsLR_den-91k_bold.dtseries.nii'))
    2.0

    """
```

File: niworkflows/viz/plots.py (hunk 3 above)
Comment (mgxd) [reply niworkflows/viz/plots.py:52]: it does - sure.

File: niworkflows/viz/plots.py (hunk 2 above)
Comment (mgxd) [reply niworkflows/viz/plots.py:33]: just reformatting

**Comment by mgxd** [comment]
decimating the data fixed the firefox rendering issue - any lingering concerns?

**Pull request merged**